# Семинар 6. Множества, словари

<hr>

### Вспомним основные операции с множествами

![](https://www.learnbyexample.org/wp-content/uploads/python/Python-Set-Operatioons.png)

<hr>

    a | b = union
    a & b = intersection
    a - b = difference
    a ^ b = symmetric difference

## Задача 1. Свадьба

Вам поручили проанализировать список гостей на свадьбе. В качестве данных вам предоставили два множества: гости со стороны жениха (groom) и гости со стороны невесты (bride). При этом некоторые гости есть в списках и жениха, и невесты одновременно.

In [ ]:
groom = {'Татьяна', 'Мария', 'Ирина', 'Юлия', 'Александр', 'Михаил', 'Максим',
         'Лев', 'Артём', 'Марк', 'Иван', 'Дмитрий', 'Матвей', 'Даниил'}
bride = {'Анастасия', 'Елена', 'Даниил', 'Ольга', 'Наталья', 'Екатерина',
         'Анна', 'Татьяна', 'Мария', 'Ирина', 'Юлия', 'Марк', 'Иван'}

Какое количество гостей приглашено на свадьбу (помните, что некоторые гости есть в списках и жениха, и невесты одновременно)?

In [ ]:
# ваш код

Выведите имена общих гостей жениха и невесты, отсортированные по алфавиту:

In [ ]:
# ваш код

Выведите имена тех гостей, которые не являются общими гостями жениха и невесты, отсортированные по алфавиту

In [ ]:
# ваш код

Выведите имена гостей невесты, но только тех, кто не являются при этом гостями жениха, отсортированные по алфавиту

In [ ]:
# ваш код

<hr>

### Вспомним работу со словарями

`py_dict = {'name': 'Bob', 'age': 25, 'job': 'Dev', 'city': 'New York', 'email': 'bob@web.com'}`

![](https://www.learnbyexample.org/wp-content/uploads/python/Dictionary-Key-Value-Pairs-Illustration.png)

<hr>

## Задача 2. Числа в словаре

Дан словарь с числовыми значениями. Необходимо их все перемножить и вывести на экран результат

In [ ]:
my_dictionary = {'data1': '375', 'data2': '567.2', 'data3': '-37', 'data4': '21'}

# ваш код


## Задача 3. Частотный анализ стихотворения

Создайте частотный словарь слов стихотворения Константина Симонова.

In [ ]:
poetry = '''Жди меня, и я вернусь.
Только очень жди,
Жди, когда наводят грусть
Желтые дожди,
Жди, когда снега метут,
Жди, когда жара,
Жди, когда других не ждут,
Позабыв вчера.
Жди, когда из дальних мест
Писем не придет,
Жди, когда уж надоест
Всем, кто вместе ждет.
Жди меня, и я вернусь,
Не желай добра
Всем, кто знает наизусть,
Что забыть пора.
Пусть поверят сын и мать
В то, что нет меня,
Пусть друзья устанут ждать,
Сядут у огня,
Выпьют горькое вино
На помин души…
Жди. И с ними заодно
Выпить не спеши.
Жди меня, и я вернусь,
Всем смертям назло.
Кто не ждал меня, тот пусть
Скажет: — Повезло.
Не понять, не ждавшим им,
Как среди огня
Ожиданием своим
Ты спасла меня.
Как я выжил, будем знать
Только мы с тобой, —
Просто ты умела ждать,
Как никто другой.'''

In [ ]:
poetry

'Жди меня, и я вернусь.\nТолько очень жди,\nЖди, когда наводят грусть\nЖелтые дожди,\nЖди, когда снега метут,\nЖди, когда жара,\nЖди, когда других не ждут,\nПозабыв вчера.\nЖди, когда из дальних мест\nПисем не придет,\nЖди, когда уж надоест\nВсем, кто вместе ждет.\nЖди меня, и я вернусь,\nНе желай добра\nВсем, кто знает наизусть,\nЧто забыть пора.\nПусть поверят сын и мать\nВ то, что нет меня,\nПусть друзья устанут ждать,\nСядут у огня,\nВыпьют горькое вино\nНа помин души…\nЖди. И с ними заодно\nВыпить не спеши.\nЖди меня, и я вернусь,\nВсем смертям назло.\nКто не ждал меня, тот пусть\nСкажет: — Повезло.\nНе понять, не ждавшим им,\nКак среди огня\nОжиданием своим\nТы спасла меня.\nКак я выжил, будем знать\nТолько мы с тобой, —\nПросто ты умела ждать,\nКак никто другой.'

In [ ]:
punct = ',.:–…—' # создаем строку со всеми знаками препинания, которые хотим удалить из текста

for p in punct:                    # проходимся циклом по знакам препинания
    poetry = poetry.replace(p, '') # и каждый заменяем на пустоту в стихотворении

poetry = poetry.replace('\n', ' ') # в конце удаляем все переносы строк

In [ ]:
poetry # получаем сырую строку

'Жди меня и я вернусь Только очень жди Жди когда наводят грусть Желтые дожди Жди когда снега метут Жди когда жара Жди когда других не ждут Позабыв вчера Жди когда из дальних мест Писем не придет Жди когда уж надоест Всем кто вместе ждет Жди меня и я вернусь Не желай добра Всем кто знает наизусть Что забыть пора Пусть поверят сын и мать В то что нет меня Пусть друзья устанут ждать Сядут у огня Выпьют горькое вино На помин души Жди И с ними заодно Выпить не спеши Жди меня и я вернусь Всем смертям назло Кто не ждал меня тот пусть Скажет  Повезло Не понять не ждавшим им Как среди огня Ожиданием своим Ты спасла меня Как я выжил будем знать Только мы с тобой  Просто ты умела ждать Как никто другой'

Приведите все слова в тексте к нижнему регистру:

In [ ]:
# ваш код
poetry = poetry.lower()

In [ ]:
poetry

'жди меня и я вернусь только очень жди жди когда наводят грусть желтые дожди жди когда снега метут жди когда жара жди когда других не ждут позабыв вчера жди когда из дальних мест писем не придет жди когда уж надоест всем кто вместе ждет жди меня и я вернусь не желай добра всем кто знает наизусть что забыть пора пусть поверят сын и мать в то что нет меня пусть друзья устанут ждать сядут у огня выпьют горькое вино на помин души жди и с ними заодно выпить не спеши жди меня и я вернусь всем смертям назло кто не ждал меня тот пусть скажет  повезло не понять не ждавшим им как среди огня ожиданием своим ты спасла меня как я выжил будем знать только мы с тобой  просто ты умела ждать как никто другой'

Создайте частотный словарь:

In [ ]:
poetry_lst = poetry.split() # для начала создадим список
print(poetry_lst)

['жди', 'меня', 'и', 'я', 'вернусь', 'только', 'очень', 'жди', 'жди', 'когда', 'наводят', 'грусть', 'желтые', 'дожди', 'жди', 'когда', 'снега', 'метут', 'жди', 'когда', 'жара', 'жди', 'когда', 'других', 'не', 'ждут', 'позабыв', 'вчера', 'жди', 'когда', 'из', 'дальних', 'мест', 'писем', 'не', 'придет', 'жди', 'когда', 'уж', 'надоест', 'всем', 'кто', 'вместе', 'ждет', 'жди', 'меня', 'и', 'я', 'вернусь', 'не', 'желай', 'добра', 'всем', 'кто', 'знает', 'наизусть', 'что', 'забыть', 'пора', 'пусть', 'поверят', 'сын', 'и', 'мать', 'в', 'то', 'что', 'нет', 'меня', 'пусть', 'друзья', 'устанут', 'ждать', 'сядут', 'у', 'огня', 'выпьют', 'горькое', 'вино', 'на', 'помин', 'души', 'жди', 'и', 'с', 'ними', 'заодно', 'выпить', 'не', 'спеши', 'жди', 'меня', 'и', 'я', 'вернусь', 'всем', 'смертям', 'назло', 'кто', 'не', 'ждал', 'меня', 'тот', 'пусть', 'скажет', 'повезло', 'не', 'понять', 'не', 'ждавшим', 'им', 'как', 'среди', 'огня', 'ожиданием', 'своим', 'ты', 'спасла', 'меня', 'как', 'я', 'выжил', 'буд

In [ ]:
a = {}
a['key'] = 1

In [ ]:
a

{'key': 1}

In [ ]:
a['key'] += 1
a

{'key': 2}

In [ ]:
poetry_dict = {}

# ваш код

for p in poetry_lst:
    if p not in poetry_dict:
        poetry_dict[p] = 1
    else:
        poetry_dict[p] += 1

In [ ]:
print(poetry_dict)

{'жди': 11, 'меня': 6, 'и': 5, 'я': 4, 'вернусь': 3, 'только': 2, 'очень': 1, 'когда': 6, 'наводят': 1, 'грусть': 1, 'желтые': 1, 'дожди': 1, 'снега': 1, 'метут': 1, 'жара': 1, 'других': 1, 'не': 7, 'ждут': 1, 'позабыв': 1, 'вчера': 1, 'из': 1, 'дальних': 1, 'мест': 1, 'писем': 1, 'придет': 1, 'уж': 1, 'надоест': 1, 'всем': 3, 'кто': 3, 'вместе': 1, 'ждет': 1, 'желай': 1, 'добра': 1, 'знает': 1, 'наизусть': 1, 'что': 2, 'забыть': 1, 'пора': 1, 'пусть': 3, 'поверят': 1, 'сын': 1, 'мать': 1, 'в': 1, 'то': 1, 'нет': 1, 'друзья': 1, 'устанут': 1, 'ждать': 2, 'сядут': 1, 'у': 1, 'огня': 2, 'выпьют': 1, 'горькое': 1, 'вино': 1, 'на': 1, 'помин': 1, 'души': 1, 'с': 2, 'ними': 1, 'заодно': 1, 'выпить': 1, 'спеши': 1, 'смертям': 1, 'назло': 1, 'ждал': 1, 'тот': 1, 'скажет': 1, 'повезло': 1, 'понять': 1, 'ждавшим': 1, 'им': 1, 'как': 3, 'среди': 1, 'ожиданием': 1, 'своим': 1, 'ты': 2, 'спасла': 1, 'выжил': 1, 'будем': 1, 'знать': 1, 'мы': 1, 'тобой': 1, 'просто': 1, 'умела': 1, 'никто': 1, 'друг

In [ ]:
for k, v in sorted(poetry_dict.items()):
    print(k, v)

будем 1
в 1
вернусь 3
вино 1
вместе 1
всем 3
вчера 1
выжил 1
выпить 1
выпьют 1
горькое 1
грусть 1
дальних 1
добра 1
дожди 1
других 1
другой 1
друзья 1
души 1
жара 1
ждавшим 1
ждал 1
ждать 2
ждет 1
жди 11
ждут 1
желай 1
желтые 1
забыть 1
заодно 1
знает 1
знать 1
и 5
из 1
им 1
как 3
когда 6
кто 3
мать 1
меня 6
мест 1
метут 1
мы 1
на 1
наводят 1
надоест 1
назло 1
наизусть 1
не 7
нет 1
никто 1
ними 1
огня 2
ожиданием 1
очень 1
писем 1
повезло 1
поверят 1
позабыв 1
помин 1
понять 1
пора 1
придет 1
просто 1
пусть 3
с 2
своим 1
скажет 1
смертям 1
снега 1
спасла 1
спеши 1
среди 1
сын 1
сядут 1
то 1
тобой 1
только 2
тот 1
ты 2
у 1
уж 1
умела 1
устанут 1
что 2
я 4


In [ ]:
new_lst = []
for k, v in poetry_dict.items():
    new_lst.append((v, k))

In [ ]:
new_lst

[(11, 'жди'),
 (6, 'меня'),
 (5, 'и'),
 (4, 'я'),
 (3, 'вернусь'),
 (2, 'только'),
 (1, 'очень'),
 (6, 'когда'),
 (1, 'наводят'),
 (1, 'грусть'),
 (1, 'желтые'),
 (1, 'дожди'),
 (1, 'снега'),
 (1, 'метут'),
 (1, 'жара'),
 (1, 'других'),
 (7, 'не'),
 (1, 'ждут'),
 (1, 'позабыв'),
 (1, 'вчера'),
 (1, 'из'),
 (1, 'дальних'),
 (1, 'мест'),
 (1, 'писем'),
 (1, 'придет'),
 (1, 'уж'),
 (1, 'надоест'),
 (3, 'всем'),
 (3, 'кто'),
 (1, 'вместе'),
 (1, 'ждет'),
 (1, 'желай'),
 (1, 'добра'),
 (1, 'знает'),
 (1, 'наизусть'),
 (2, 'что'),
 (1, 'забыть'),
 (1, 'пора'),
 (3, 'пусть'),
 (1, 'поверят'),
 (1, 'сын'),
 (1, 'мать'),
 (1, 'в'),
 (1, 'то'),
 (1, 'нет'),
 (1, 'друзья'),
 (1, 'устанут'),
 (2, 'ждать'),
 (1, 'сядут'),
 (1, 'у'),
 (2, 'огня'),
 (1, 'выпьют'),
 (1, 'горькое'),
 (1, 'вино'),
 (1, 'на'),
 (1, 'помин'),
 (1, 'души'),
 (2, 'с'),
 (1, 'ними'),
 (1, 'заодно'),
 (1, 'выпить'),
 (1, 'спеши'),
 (1, 'смертям'),
 (1, 'назло'),
 (1, 'ждал'),
 (1, 'тот'),
 (1, 'скажет'),
 (1, 'повезло'),
 (1,

In [ ]:
print(sorted(new_lst, reverse=True))

[(11, 'жди'), (7, 'не'), (6, 'меня'), (6, 'когда'), (5, 'и'), (4, 'я'), (3, 'пусть'), (3, 'кто'), (3, 'как'), (3, 'всем'), (3, 'вернусь'), (2, 'что'), (2, 'ты'), (2, 'только'), (2, 'с'), (2, 'огня'), (2, 'ждать'), (1, 'устанут'), (1, 'умела'), (1, 'уж'), (1, 'у'), (1, 'тот'), (1, 'тобой'), (1, 'то'), (1, 'сядут'), (1, 'сын'), (1, 'среди'), (1, 'спеши'), (1, 'спасла'), (1, 'снега'), (1, 'смертям'), (1, 'скажет'), (1, 'своим'), (1, 'просто'), (1, 'придет'), (1, 'пора'), (1, 'понять'), (1, 'помин'), (1, 'позабыв'), (1, 'поверят'), (1, 'повезло'), (1, 'писем'), (1, 'очень'), (1, 'ожиданием'), (1, 'ними'), (1, 'никто'), (1, 'нет'), (1, 'наизусть'), (1, 'назло'), (1, 'надоест'), (1, 'наводят'), (1, 'на'), (1, 'мы'), (1, 'метут'), (1, 'мест'), (1, 'мать'), (1, 'им'), (1, 'из'), (1, 'знать'), (1, 'знает'), (1, 'заодно'), (1, 'забыть'), (1, 'желтые'), (1, 'желай'), (1, 'ждут'), (1, 'ждет'), (1, 'ждал'), (1, 'ждавшим'), (1, 'жара'), (1, 'души'), (1, 'друзья'), (1, 'другой'), (1, 'других'), (1, '

<hr>

<center><h3>Профессиональная часть (не подготовка к НЭ) – Natural Language Processing </h3> </center>

В частотном анализе выше мы видим, что слова в разных склонениях считаются как разные. Это проблема лемматизации – то есть слова не приведены к исходной форме (так, как они записаны в словаре).

Импортируем библиотеку pymorphy3 – морфологический анализатор для русского языка, написанный на языке Python и использующий словари из [OpenCorpora](https://opencorpora.org/dict.php).

Используем оттуда MorphAnalyzer. Для заданного слова он может найти все возможные формы.

In [ ]:
!pip install pymorphy3 # для начала загрузим библиотеку

<blockquote>Библиотека — это набор связанных модулей, объединенных вместе для решения определенной задачи или предоставления определенного функционала. Библиотеки можно представить как расширение базового функционала Python, которое помогает разработчикам решать различные задачи без необходимости писать все с нуля.</blockquote>

In [ ]:
import pymorphy3 # импортируем библиотеку в нашу тетрадку
morph = pymorphy3.MorphAnalyzer() # создаем морфологический анализатор

In [ ]:
morph.parse('жди') # используем метод morph.parse('слово'), чтобы найти исходную форму

[Parse(word='жди', tag=OpencorporaTag('VERB,impf,tran sing,impr,excl'), normal_form='ждать', score=1.0, methods_stack=((DictionaryAnalyzer(), 'жди', 973, 11),))]

Разберем, что мы получили:

`word = 'жди'` – это введенное нами слова для анализа.

`tag=OpencorporaTag('VERB,impf,tran sing,impr,excl')` – это набор граммем, характеризующих данное слово.

`VERB` – глагол \
`impf` – несовершенный вид \
`tran sing` – переходный глагол в единственном числе \
`impr` – повелительное наклонение \
`excl` – говорящий не включён в действие (жди, ждите)

[Полный список граммем](https://opencorpora.org/dict.php?act=gram)

`normal_form='ждать'` – это каноническая форма слова (например, форма единственного числа, именительного падежа для существительных).

`score=1.0` – это оценка вероятности того, что данный разбор правильный.

`methods_stack=((DictionaryAnalyzer(), 'жди', 973, 11)` – техническая информация о том, как он нашел слово

In [ ]:
morph.parse('жди')[0].normal_form # [0] – заходим внутрь списка, забираем нормальную форму

'ждать'

Осталось только пропустить наше стихотворение через анализатор.

In [ ]:
poetry_dict = {}

for word in poetry_lst:
    word = morph.parse(word)[0].normal_form
    if word not in poetry_dict:
        poetry_dict[word] = 1
    else:
        poetry_dict[word] += 1

In [ ]:
print(poetry_dict)

{'ждать': 17, 'я': 10, 'и': 5, 'вернуться': 3, 'только': 2, 'очень': 1, 'когда': 6, 'наводить': 1, 'грусть': 1, 'жёлтый': 1, 'дождь': 1, 'снег': 1, 'мести': 1, 'жара': 1, 'другой': 2, 'не': 7, 'позабыть': 1, 'вчера': 1, 'из': 1, 'дальний': 1, 'место': 1, 'письмо': 1, 'прийти': 1, 'уж': 1, 'надоесть': 1, 'весь': 3, 'кто': 3, 'вместе': 1, 'желать': 1, 'добро': 1, 'знать': 2, 'наизусть': 1, 'что': 2, 'забыть': 1, 'пора': 1, 'пусть': 3, 'поверить': 1, 'сын': 1, 'мать': 1, 'в': 1, 'то': 1, 'нет': 1, 'друг': 1, 'устать': 1, 'сесть': 1, 'у': 1, 'огонь': 2, 'выпить': 2, 'горький': 1, 'вино': 1, 'на': 1, 'помин': 1, 'душа': 1, 'с': 2, 'они': 2, 'заодно': 1, 'спешить': 1, 'смерть': 1, 'назло': 1, 'тот': 1, 'сказать': 1, 'повезти': 1, 'понять': 1, 'как': 3, 'среди': 1, 'ожидание': 1, 'свой': 1, 'ты': 3, 'спасти': 1, 'выжить': 1, 'быть': 1, 'мы': 1, 'просто': 1, 'уметь': 1, 'никто': 1}


Согласитесь, что теперь частотный анализ выглядит гораздо лучше! А наш изначальный код изменился всего на одну строку!

И еще один небольшой лайфхак – как быстро делать частотный анализ текста: библиотека `collections`.

In [ ]:
from collections import Counter # импортируем "частотный анализатор"

In [ ]:
for i in range(len(poetry_lst)):
    poetry_lst[i] = morph.parse(poetry_lst[i])[0].normal_form # изменяем наш список

In [ ]:
print(poetry_lst)

['ждать', 'я', 'и', 'я', 'вернуться', 'только', 'очень', 'ждать', 'ждать', 'когда', 'наводить', 'грусть', 'жёлтый', 'дождь', 'ждать', 'когда', 'снег', 'мести', 'ждать', 'когда', 'жара', 'ждать', 'когда', 'другой', 'не', 'ждать', 'позабыть', 'вчера', 'ждать', 'когда', 'из', 'дальний', 'место', 'письмо', 'не', 'прийти', 'ждать', 'когда', 'уж', 'надоесть', 'весь', 'кто', 'вместе', 'ждать', 'ждать', 'я', 'и', 'я', 'вернуться', 'не', 'желать', 'добро', 'весь', 'кто', 'знать', 'наизусть', 'что', 'забыть', 'пора', 'пусть', 'поверить', 'сын', 'и', 'мать', 'в', 'то', 'что', 'нет', 'я', 'пусть', 'друг', 'устать', 'ждать', 'сесть', 'у', 'огонь', 'выпить', 'горький', 'вино', 'на', 'помин', 'душа', 'ждать', 'и', 'с', 'они', 'заодно', 'выпить', 'не', 'спешить', 'ждать', 'я', 'и', 'я', 'вернуться', 'весь', 'смерть', 'назло', 'кто', 'не', 'ждать', 'я', 'тот', 'пусть', 'сказать', 'повезти', 'не', 'понять', 'не', 'ждать', 'они', 'как', 'среди', 'огонь', 'ожидание', 'свой', 'ты', 'спасти', 'я', 'как', 'я

In [ ]:
Counter(poetry_lst).most_common(20) # печатаем

[('ждать', 17),
 ('я', 10),
 ('не', 7),
 ('когда', 6),
 ('и', 5),
 ('вернуться', 3),
 ('весь', 3),
 ('кто', 3),
 ('пусть', 3),
 ('как', 3),
 ('ты', 3),
 ('только', 2),
 ('другой', 2),
 ('знать', 2),
 ('что', 2),
 ('огонь', 2),
 ('выпить', 2),
 ('с', 2),
 ('они', 2),
 ('очень', 1)]

<center><h3>Конец профессиональной части</h3> </center>

<hr>

## A6 НЭ

В переменных tea и coffee хранятся множества строк с именами респондентов, которые любят чай и кофе соответственно.

Выберите все выражения (минимум два), которые помогут найти всех, кто любит оба напитка, и сохранить результат в переменную tea_and_coffee.

* tea_and_coffee = coffee.intersection(tea)

* tea_and_coffee = tea.symmetric_intersection(tea_and_coffee)

* tea_and_coffee = tea & coffee

* tea_and_coffee = (tea | coffee) - (tea ^ coffee)

* tea_and_coffee = tea.intersection(coffee)

In [ ]:
# проверка


## A6 НЭ

В переменных products и recipe хранятся множества строк с названиями продуктов, которые есть в ходильнике и используются в рецепте блюда соответственно.

Выберите все выражения (минимум два), которые помогут определить, каких продуктов не хватает для приготовления блюда, и сохранить результат в переменную to_buy.


* to_buy = products.difference(recipe)

* to_buy = products.symmetric_difference(recipe)

* to_buy = recipe.difference(products)

* to_buy = recipe - products

* to_buy = products ^ recipe

In [ ]:
# проверка


## B5 НЭ

Реализуйте программу, которая поможет Ефросинье выбрать IT-компанию для стажировки.

* В переменную companies_top считывается список компаний, которые вошли в топ-10 работодателей. Этот код уже написан.
* В переменную сompanies_python считывается список компаний, в которых работать надо будет с использованием языка Python. Этот код уже написан.
* В переменную companies_money считывается список компаний, в которых стажировка оплачивается. Этот код уже написан.
* Гарантируется, что названия компаний в каждом из трех списков уникальные и что в каждом списке есть как минимум одна компания.
* Выведите через запятую и пробел в алфавитном порядке названия компаний с оплачиваемой стажировкой, которые при этом не входят в топ-10 работодателей, и в которых надо будет работать с использованием Python.

**Пример** \
**ВВОД**

ВК, Сбербанк \
ПИК, Тинькофф, Сбербанк \
Тинькофф, ПИК, ВК, Перекресток

**ВЫВОД**

ПИК, Тинькофф

In [ ]:
companies_top = input().split(', ')
companies_python = input().split(', ')
companies_money = input().split(', ')

# ваш код ниже


## A7 НЭ

В словаре lang_group хранятся пары объектов, в которых ключи — названия языков, а значения — языковые группы.

Программа ниже должна печатать True, если вторая буква названия языка и вторая буква названия его языковой группы совпадают, и False в ином случае.

`lang_group = {
    'Немецкий': 'Германская',
    'Литовский': 'Балтийская',
    'Сербский': 'Славянская'
}`

`for lang, group in lang_group.items():
    print(__________)`
    
Выберите такой вариант из выпадающего списка , чтобы программа работала корректно.

* `lang[2] == group[2]`
* `lang[1] == lang[group][1]`
* `lang == group`
* `lang[1] == group[1]`
* `lang[1] != group[1]`

In [ ]:
lang_group = {
    'Немецкий': 'Германская',
    'Литовский': 'Балтийская',
    'Сербский': 'Славянская'
}

for lang, group in lang_group.items():
    print(lang[1] != group[1])

False
True
True


## B6 НЭ

Реализуйте программу, которая проверяет интернет-провайдеров с точки зрения выполнения договорных условий и находит надежных провайдеров.

* Дан словарь internet, в котором ключами являются названия компаний, а значениями — списки с индикаторами скорости за последнюю неделю: \
2 — скорость превышала заявленную; \
1 — скорость соответствовала заявленной; \
0 — была ниже.
* Гарантируется, что названия компаний не повторяются.
* В переменную filter_days считывается минимально допустимое количество дней со скоростью выше заявленной. Этот код уже написан.
* Программа проверяет, что в статистике компании нет дней с понижением скорости и что количество дней с повышенной скоростью не менее значения filter_days, и выводит названия таких компаний, каждое на новой строке.
* Если таких компаний нет — программа ничего не выводит.

In [ ]:
internet = {'Мостелеком': [1, 2, 1, 1, 1, 2, 1],
            'Соединение2023': [1, 1, 0, 1, 1, 1, 0],
            'МТТС': [1, 1, 1, 1, 2, 1, 1],
            'Трилайн': [2, 2, 1, 1, 1, 2, 2]}

filter_days = int(input())

# ваш код ниже

## B6 НЭ

Реализуйте программу, которая проверяет, насколько популярна языковая школа.
* Дан словарь schools, в котором ключами являются названия школ, а значениями — списки с количеством заявок на обучение за последнюю неделю. Гарантируется, что названия школ не повторяются.
* В переменную filter_cust считывается минимально допустимое среднее количество заявок за неделю, чтобы школа не вышла в "минус". Этот код уже написан.
* Программа проверяет, что в статистике подачи заявок в школу среднее число заявок не ниже значения filter_cust, и выводит названия таких школ. Названия выводятся в том же порядке, в котором они хранятся в словаре. Каждое название выводится на новой строке.
* Если таких школ нет — программа ничего не выводит.

In [ ]:
schools = {'ЯзыкДляВсех': [45, 67, 30, 24, 82, 12, 91],
           'Питонисты2023': [32, 14, 29, 100, 21, 70, 25],
           'EngStart': [105, 48, 22, 74, 53, 90, 12],
           'NewLang': [83, 200, 0, 44, 12, 54, 19]}

filter_cust = int(input())

# ваш код ниже